### PR TITLE
cling: 0.9 -> 1.0

### PR DESCRIPTION
**Pull request by thomasjm**
## Description of changes

Working on Cling 1.0, per the advice of @vgvassilev on #283989.

This currently builds, but has some path troubles. Marking as draft for now.

CC @a-n-n-a-l-e-e , @wegank 

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thomasjm** [comment]
Okay, this seems to be working now. Ready for review!

**Review by RossComputerGuy — COMMENTED** [review]
```
$ ./result/bin/cling
In file included from input_line_1:1:
/nix/store/8y16aj4mk0rbck2hlg538322d5r1gi7m-gcc-13.2.0/include/c++/13.2.0/new:40:10: fatal error: 'bits/c++config.h' file not found
#include <bits/c++config.h>
         ^~~~~~~~~~~~~~~~~~
Warning in cling::IncrementalParser::CheckABICompatibility():
  Failed to extract C++ standard library version.
Replaced symbol atexit cannot be found in JIT!
Replaced symbol at_quick_exit cannot be found in JIT!

****************** CLING ******************
* Type C++ code and press enter to run it *
*             Type .q to exit             *
*******************************************
[cling]$
```
Builds fine on my M1 Pro under NixOS, but these errors are concerning. Though I don't know anything about cling but these errors still concern me.

File: pkgs/development/interpreters/cling/default.nix [hunk 1]
```diff
@@ -126,9 +120,9 @@ let
 
   # Runtime flags for the C++ standard library
   cxxFlags = if useLLVMLibcxx then [
-    "-I" "${lib.getDev llvmPackages_9.libcxx}/include/c++/v1"
-    "-L" "${llvmPackages_9.libcxx}/lib"
-    "-l" "${llvmPackages_9.libcxx}/lib/libc++.so"
+    "-I" "${lib.getDev llvmPackages_13.libcxx}/include/c++/v1"
+    "-L" "${llvmPackages_13.libcxx}/lib"
+    "-l" "${llvmPackages_13.libcxx}/lib/libc++.so"
```
Comment (wegank) [inline pkgs/development/interpreters/cling/default.nix:125]:
```suggestion
    "-l" "${llvmPackages_13.libcxx}/lib/libc++${stdenv.hostPlatform.extensions.sharedLibrary}"
```

**Comment by wegank** [comment]
> ```
> $ ./result/bin/cling
> In file included from input_line_1:1:
> /nix/store/8y16aj4mk0rbck2hlg538322d5r1gi7m-gcc-13.2.0/include/c++/13.2.0/new:40:10: fatal error: 'bits/c++config.h' file not found
> #include <bits/c++config.h>
>          ^~~~~~~~~~~~~~~~~~
> Warning in cling::IncrementalParser::CheckABICompatibility():
>   Failed to extract C++ standard library version.
> Replaced symbol atexit cannot be found in JIT!
> Replaced symbol at_quick_exit cannot be found in JIT!
> 
> ****************** CLING ******************
> * Type C++ code and press enter to run it *
> *             Type .q to exit             *
> *******************************************
> [cling]$
> ```
> 
> Builds fine on my M1 Pro under NixOS, but these errors are concerning. Though I don't know anything about cling but these errors still concern me.

Looks like enabling `useLLVMLibcxx` on darwin should fix the issue? Something GCC-related is always strange on darwin.

**Comment by RossComputerGuy** [comment]
> enabling `useLLVMLibcxx` on darwin

The problem is I am not under Darwin but under NixOS on Apple Silicon.

**Comment by wegank** [comment]
Oh, that's not nice at all. But maybe that will still help?

**Comment by RossComputerGuy** [comment]
Well, I get even more interesting errors with that option enabled:
```
$ ./result/bin/cling
Warning in cling::IncrementalParser::CheckABICompatibility():
  Failed to extract C++ standard library version.
In file included from input_line_3:1:
/nix/store/nrqmhzgg3wfnqf2qrqdrl6mbss3lfp9w-libcxx-13.0.1/lib/libc++.so:1:1: error: C++ requires a type specifier for all declarations
INPUT(libc++.so.1 -lc++abi)
^
/nix/store/nrqmhzgg3wfnqf2qrqdrl6mbss3lfp9w-libcxx-13.0.1/lib/libc++.so:1:7: error: use of undeclared identifier 'libc'
INPUT(libc++.so.1 -lc++abi)
      ^
/nix/store/nrqmhzgg3wfnqf2qrqdrl6mbss3lfp9w-libcxx-13.0.1/lib/libc++.so:1:28: error: expected ';' after top level declarator
INPUT(libc++.so.1 -lc++abi)
                           ^
                           ;

****************** CLING ******************
* Type C++ code and press enter to run it *
*             Type .q to exit             *
*******************************************
[cling]$
```

**Comment by wegank** [comment]
What if we replaced `x86_64-unknown-linux-gnu` with `${stdenv.hostPlatform.config}` in the expression, while keeping `useLLVMLibcxx` disabled on Linux?

**Comment by RossComputerGuy** [comment]
> So `useLLVMLibcxx` doesn't look so good on Linux.

At least on `aarch64-linux`, no it's not good at all.



> What if we replaced `x86_64-unknown-linux-gnu` with `${stdenv.hostPlatform.config}` in the expression, while keeping `useLLVMLibcxx` disabled on Linux?

Amazing, it works.
```
$ ./result/bin/cling

****************** CLING ******************
* Type C++ code and press enter to run it *
*             Type .q to exit             *
*******************************************
[cling]$
```

**Comment by wegank** [comment]
Thanks! Both changes should also be needed in 0.9, so I assume they are not regressions of this PR. I'll create a follow-up PR for that.

**Review by wegank — APPROVED** [review]

**Pull request merged**